modules/dnn/src/layers/slice_layer.cpp

```diff
@@ -154,6 +180,13 @@ class SliceLayerImpl : public SliceLayer
         CV_Assert(inputs.size() == 1);
         MatShape inpShape = inputs[0];
 
+        if (!initComplete)
+        {
+            initComplete = true;
+            SliceLayerImpl* that = const_cast<SliceLayerImpl*>(this); // No guarantee it's safe.
```

Comment:
Please, use "updateMemoryShapes" function instead of removing constness. It will be called from "updateLayersShapes" in Net::Impl class before other shapes are computed. So it should work for your case

Reply:
It is good suggestion, but `getMemoryShapes` might be called from `void Net::getLayersShapes(/* ... */) const` without previous call of `updateLayersShapes`.

Reply:
Maybe it is possible to adapt Sergey idea. Will the following might work?
- Create a copy of already filled `sliceRanges` in `getMemoryShapes`
- Add a free function:
  ```cpp
  void setAxisFromShape(const MatShape& inpShape, std::vector<std::vector<Range> >& sliceRanges) {
    /* all logic */
  } 
  ```
- Substitue `initComplete` with  `shapesInitialized` check
- Perform `sliceRanges` member initialization in `updateLayersShapes`

Reply:
@VadimLevin, done.